pkg/vm/builtins/ecdsa.go

```diff
@@ -0,0 +1,160 @@
+	}
+
+	pubKey := &ecdsa.PublicKey{A: key}
+	sig, ok := e.signatures[pubOffset]
+	if !ok {
+		return fmt.Errorf("Signature is missing form ECDA builtin")
+	}
+
+	msgBytes := msgField.Bytes()
+	valid, err := pubKey.Verify(sig.Bytes(), msgBytes[:], nil)
+	if err != nil {
+		return err
+	}
+
+	if !valid {
+		// Now try with Neg Y. Already know the point is on the curve so no need to check again
+		key = starkcurve.G1Affine{X: *pubX, Y: *negY}
+		pubKey = &ecdsa.PublicKey{A: key}
+		valid, err := pubKey.Verify(sig.Bytes(), msgBytes[:], nil)
+		if err != nil {
+			return err
+		}
+		if !valid {
+			return fmt.Errorf("Signature is not valid")
+		}
+	}
+	fmt.Println("VALID")
+	return nil
+}
+
+func (e *ECDSA) InferValue(segment *memory.Segment, offset uint64) error {
+	return fmt.Errorf("Can't infer value")
+}
+
+/*
+Hint that will call this function looks like this:
+
+	"hints": {
+	    "6": [
+	        {
+	            "accessible_scopes": [
+	                "starkware.cairo.common.signature",
+	                "starkware.cairo.common.signature.verify_ecdsa_signature"
+	            ],
+	            "code": "ecdsa_builtin.add_signature(ids.ecdsa_ptr.address_, (ids.signature_r, ids.signature_s))",
+	            "flow_tracking_data": {
+	                "ap_tracking": {
+	                    "group": 2,
+	                    "offset": 0
+	                },
+	                "reference_ids": {
+	                    "starkware.cairo.common.signature.verify_ecdsa_signature.ecdsa_ptr": 4,
+	                    "starkware.cairo.common.signature.verify_ecdsa_signature.message": 0,
+	                    "starkware.cairo.common.signature.verify_ecdsa_signature.public_key": 1,
+	                    "starkware.cairo.common.signature.verify_ecdsa_signature.signature_r": 2,
+	                    "starkware.cairo.common.signature.verify_ecdsa_signature.signature_s": 3
+	                }
+	            }
+	        }
+	    ]
+	},
+*/
+func (e *ECDSA) AddSignature(pubOffset uint64, r, s fp.Element) error {
+	if e.signatures == nil {
+		e.signatures = make(map[uint64]ecdsa.Signature)
+	}
+	bytes := make([]byte, 0, 64)
+	rBytes := r.Bytes()
+	bytes = append(bytes, rBytes[:]...)
+	sBytes := s.Bytes()
+	bytes = append(bytes, sBytes[:]...)
+
+	sig := ecdsa.Signature{}
+	_, err := sig.SetBytes(bytes)
+	if err != nil {
+		return err
+	}
+
+	e.signatures[pubOffset] = sig
+	return nil
+}
+
+func (e *ECDSA) String() string {
+	return ECDSAName
+}
+
+// recoverY recovers the y and -y coordinate of x. True y can be either y or -y
+func recoverY(x *fp.Element) (*fp.Element, *fp.Element, error) {
+	ALPHA := fp.NewElement(1)
+	BETA := fp.Element{}
+	_, _ = BETA.SetString("3141592653589793238462643383279502884197169399375105820974944592307816406665")
+	// y_squared = (x * x * x + ALPHA * x + BETA) % FIELD_PRIME
+	x2 := new(fp.Element).Mul(x, x)
+	x3 := x2.Mul(x2, x)
+	a := new(fp.Element).Mul(&ALPHA, x)
+	x3.Add(x3, a)
+	x3.Add(x3, &BETA)
+	y := x3.Sqrt(x3)
```

Comment:
no need to store a new `y` as well, you could reuse `x3`. I understand that this could feel restrictive and not good practice, but this is code that should be as fast/cheap as possible. You could aid yourself from comments to mitigate this

Reply:
Here we store y because `sqrt` returns nil if the square root doesn't exist. Otherwise we would need check if x2 changed which would require another variable either way